leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/QueuedRequestSender.java

```diff
@@ -0,0 +1,282 @@
+import org.eclipse.leshan.core.observation.Observation;
+import org.eclipse.leshan.core.request.BindingMode;
+import org.eclipse.leshan.core.request.DownlinkRequest;
+import org.eclipse.leshan.core.response.ErrorCallback;
+import org.eclipse.leshan.core.response.LwM2mResponse;
+import org.eclipse.leshan.core.response.ResponseCallback;
+import org.eclipse.leshan.server.Stoppable;
+import org.eclipse.leshan.server.client.Client;
+import org.eclipse.leshan.server.client.ClientRegistry;
+import org.eclipse.leshan.server.client.ClientRegistryListener;
+import org.eclipse.leshan.server.client.ClientUpdate;
+import org.eclipse.leshan.server.observation.ObservationRegistry;
+import org.eclipse.leshan.server.observation.ObservationRegistryListener;
+import org.eclipse.leshan.server.queue.MessageStore;
+import org.eclipse.leshan.server.queue.QueuedRequest;
+import org.eclipse.leshan.server.queue.QueuedRequestFactory;
+import org.eclipse.leshan.server.request.LwM2mRequestSender;
+import org.eclipse.leshan.util.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This sender is a special implementation of a {@link LwM2mRequestSender}, which is aware of a LwM2M client's binding
+ * mode. If the client supports "Q" (queue) mode, then the request is processed using the internal queue, i.e. the
+ * request is enqueued first and then sent if a client is back online. If the client does not support the queue mode in
+ * its binding, then the delegate sender is called in order to send the request.
+ */
+public class QueuedRequestSender implements LwM2mRequestSender, Stoppable {
+    private static final Logger LOG = LoggerFactory.getLogger(QueuedRequestSender.class);
+    private final LwM2mRequestSender delegateSender;
+    private final ExecutorService processingExecutor = new ExceptionAwareExecutorService(
+            Executors.newCachedThreadPool(new NamedThreadFactory("leshan-qmode-processingExecutor-%d")));
+    private final Map<Long, ResponseContext> responseContextHolder = new ConcurrentHashMap<>();
+    private final MessageStore messageStore;
+    private final QueuedRequestFactory queuedRequestFactory;
+    private final QueueModeClientRegistryListener queueModeClientRegistryListener;
+    private final QueueModeObservationRegistryListener queueModeObservationRegistryListener;
+    private final ClientRegistry clientRegistry;
+    private final ObservationRegistry observationRegistry;
+    private final ClientStatusTracker clientStatusTracker;
+
+    /**
+     * Creates a new QueueRequestSender using given builder.
+     *
+     * @param builder sender builder
+     */
+    private QueuedRequestSender(Builder builder) {
+        this.messageStore = builder.messageStore;
+        this.queuedRequestFactory = builder.queuedRequestFactory;
+        this.clientRegistry = builder.clientRegistry;
+        this.observationRegistry = builder.observationRegistry;
+        this.delegateSender = builder.delegateSender;
```

Comment:
I still don't get it. The class has no setters, does it? Its state is only changed by means of the its other methods which are there regardless of whether the fields are defined final or not.